cmd/frontend/graphqlbackend/code_hosts_test.go

```diff
@@ -0,0 +1,290 @@
+package graphqlbackend
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"testing"
+
+	mockassert "github.com/derision-test/go-mockgen/testutil/assert"
+	"github.com/sourcegraph/sourcegraph/internal/database"
+	"github.com/sourcegraph/sourcegraph/internal/database/dbmocks"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc"
+	"github.com/sourcegraph/sourcegraph/internal/types"
+	"github.com/sourcegraph/sourcegraph/lib/pointers"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestSchemaResolver_CodeHosts(t *testing.T) {
+	t.Parallel()
+
+	testCodeHosts := []*types.CodeHost{
+		{
+			ID:                          1,
+			URL:                         "github.com",
+			Kind:                        extsvc.KindGitHub,
+			APIRateLimitQuota:           pointers.Ptr(int32(1)),
+			APIRateLimitIntervalSeconds: pointers.Ptr(int32(1)),
+			GitRateLimitIntervalSeconds: pointers.Ptr(int32(1)),
+			GitRateLimitQuota:           pointers.Ptr(int32(1)),
+		},
+		{
+			ID:                          2,
+			URL:                         "gitlab.com",
+			Kind:                        extsvc.KindGitLab,
+			APIRateLimitQuota:           pointers.Ptr(int32(2)),
+			APIRateLimitIntervalSeconds: pointers.Ptr(int32(2)),
+			GitRateLimitIntervalSeconds: pointers.Ptr(int32(2)),
+			GitRateLimitQuota:           pointers.Ptr(int32(2)),
+		},
+		{
+			ID:   3,
+			URL:  "bitbucket-cloud.com",
+			Kind: extsvc.KindBitbucketCloud,
+		},
+		{
+			ID:                          4,
+			URL:                         "bitbucket.com",
+			Kind:                        extsvc.KindBitbucketServer,
+			APIRateLimitQuota:           pointers.Ptr(int32(4)),
+			APIRateLimitIntervalSeconds: pointers.Ptr(int32(4)),
+			GitRateLimitIntervalSeconds: pointers.Ptr(int32(4)),
+			GitRateLimitQuota:           pointers.Ptr(int32(4)),
+		},
+	}
```

Comment:
Nitpick, but I think the `pointers.Ptr` usage is a bit overdone :)

```suggestion
	newCodeHost := func(id int32, url, kind string, quota int32) *types.CodeHost {
		var q *int32 = nil
		if quota != 0 {
			q = &quota
		}
		return &types.CodeHost{
			ID:                          id,
			URL:                         url,
			Kind:                        kind,
			APIRateLimitQuota:           q,
			APIRateLimitIntervalSeconds: q,
			GitRateLimitQuota:           q,
			GitRateLimitIntervalSeconds: q,
		}
	}

	testCodeHosts := []*types.CodeHost{
		newCodeHost(1, "github.com", extsvc.KindGitHub, 1),
		newCodeHost(2, "gitlab.com", extsvc.KindGitLab, 2),
		newCodeHost(3, "bitbucket-cloud.com", extsvc.KindBitbucketCloud, 0),
		newCodeHost(4, "bitbucket-cloud.com", extsvc.KindBitbucketCloud, 4),
	}
```